src/quantization/sq4_quantizer.h

```diff
@@ -0,0 +1,198 @@
+    this->code_size_ = vector_encoded_size_;
+}
+
+template <MetricType Metric>
+bool
+SQ4Quantizer<Metric>::TrainImpl(const DataType* data, uint64_t count) {
+    if (data == nullptr) {
+        return false;
+    }
+
+    for (uint32_t i = 0; i < count; i++) {
+        for (uint32_t d = 0; d < this->dim_; d++) {
+            auto val = data[i * this->dim_ + d];
+            if (val > diff_[d]) {
+                diff_[d] = val;
+            }
+            if (val < lower_bound_[d]) {
+                lower_bound_[d] = val;
+            }
+        }
+    }
+
+    for (uint32_t d = 0; d < this->dim_; d++) {
+        diff_[d] -= lower_bound_[d];
+    }
+
+    this->is_trained_ = true;
+    return true;
+}
+
+template <MetricType Metric>
+bool
+SQ4Quantizer<Metric>::EncodeOneImpl(const DataType* data, uint8_t* codes) const {
+    float delta;
+    uint8_t scaled;
+
+    for (uint32_t d = 0; d < this->dim_; d++) {
+        delta = ((data[d] - lower_bound_[d]) / diff_[d]);
+        if (delta < 0.0) {
+            delta = 0;
+        }
+        if (delta > 0.999) {
+            delta = 1;
+        }
+        scaled = 15 * delta;
+
+        if (d & 1) {
+            codes[d / 2] |= scaled << 4;
+        } else {
+            codes[d / 2] = 0;
+            codes[d / 2] |= scaled;
+        }
+    }
+
+    return true;
+}
+
+template <MetricType Metric>
+bool
+SQ4Quantizer<Metric>::EncodeBatchImpl(const DataType* data, uint8_t* codes, uint64_t count) {
+    for (uint64_t i = 0; i < count; ++i) {
+        this->EncodeOneImpl(data + i * this->dim_, codes + i * this->code_size_);
+    }
+    return true;
+}
+
+template <MetricType Metric>
+bool
+SQ4Quantizer<Metric>::DecodeOneImpl(const uint8_t* codes, DataType* data) {
+    for (uint32_t d = 0; d < this->dim_; d++) {
+        if (d & 1) {
+            data[d] = ((codes[d / 2] & 0xf0) >> 4) / 15.0 * diff_[d] + lower_bound_[d];
+            data[d] = data[d];
+        } else {
+            data[d] = (codes[d / 2] & 0x0f) / 15.0 * diff_[d] + lower_bound_[d];
+            data[d] = data[d];
+        }
+    }
+
+    return true;
+}
+
+template <MetricType Metric>
+bool
+SQ4Quantizer<Metric>::DecodeBatchImpl(const uint8_t* codes, DataType* data, uint64_t count) {
+    for (uint64_t i = 0; i < count; ++i) {
+        this->DecodeOneImpl(codes + i * this->code_size_, data + i * this->dim_);
+    }
+    return true;
+}
+
+template <MetricType Metric>
+inline float
+SQ4Quantizer<Metric>::ComputeImpl(const uint8_t* codes1, const uint8_t* codes2) {
+    if (Metric == MetricType::METRIC_TYPE_L2SQR) {
+        return SQ4ComputeCodesL2Sqr(codes1, codes2, lower_bound_.data(), diff_.data(), this->dim_);
+    } else if (Metric == MetricType::METRIC_TYPE_IP) {
+        return SQ4ComputeCodesIP(codes1, codes2, lower_bound_.data(), diff_.data(), this->dim_);
+    } else if (Metric == MetricType::METRIC_TYPE_COSINE) {
+        return SQ4ComputeCodesIP(codes1, codes2, lower_bound_.data(), diff_.data(), this->dim_);
```

Comment:
There are not implement the `Cosine Metric` ?

Reply:
yes, for now, we use IP as a temporary implement for cosine.